### ISSUE TITLE
Train new model for spinal cord segmentation

### ISSUE THREAD

**Issue by jcohenadad** [ivadomed/ivadomed#495]
### Context

SCT's current most used function for spinal cord segmentation is `sct_deepseg_sc`. With the recent introduction of `sct_deepseg`, we would like to create a new spinal cord segmentation model that would ultimately replace `sct_deepseg_sc`.

### Motivations
- `sct_deepseg` will replace all deep learning functions in SCT, facilitating the maintenance of the other individual CLIs and specific APIs.
- This replacement will enable us to get rid of TensorFlow dependency (given that `sct_deepseg` depends on Pytorch)
- `sct_deepseg` will leverage `ivadomed` for the following tasks:
  - Image pre-processing (resampling, splitting into patches, etc.)
  - Image post-processing (fill holes, threshold, small object removal, etc.)

### Specifications
- A current limitation of `sct_deepseg_sc` is that a specific model was trained for each contrast (T1, T2, T2*, DWI). We would like to replace this approach, and have a single model only, trained with all contrasts. Several possibilities to explore:
  - Pool all the contrasts and train a single-model using U-Net architecture
  - Do what is listed in the previous point, with the FiLM architecture, where the contrast is used to modulate FiLM layers
  - Do what is listed in the previous point, with the HeMIS architecture, allowing to input multiple contrasts at once (possibly more robust), and allowing for missing contrasts. This point is under investigation by @leandervaneekelen (https://github.com/ivadomed/ivadomed/projects/3)
- Another limitation is that the segmentation result varies across contrasts. For example, the segmentation is smaller on the T1 vs. on the T2, as discussed in this [forum post](https://forum.spinalcordmri.org/t/strange-phenomenon-when-using-sct-process-segmentation-to-compute-the-csa/477).
- We could consider joining the efforts of @leandervaneekelen and making it a multi-class segmentation project (cord & CSF).
- We could consider using a softseg training, which would potentially provide a more accurate computation of CSA down the line. 
- Implement an uncertainty feature. 
- Make sure to include a variety of "bad" datasets (artifacts)
- Make sure to include data with pathologies (MS, compression, etc.)

### List of tasks

- Identify the datasets to consider
  - https://github.com/spine-generic/data-multi-subject (public)
  - sct_testing/large (private)
- Organize the datasets with git/git-annex so the SHA can be incorporated in the training script, so that the model can be reproduced. 
- Check ground truth segmentations
  - Especially: update the ones for the T1w contrast (under-segmented). 
  - We adopt a "deep active learning" strategy: start training a "good" model with little data, then apply it to a bigger pool, then correct segmentations in a few data, then re-train with bigger pool, etc. 
- Establish a procedure for co-registering all existing contrasts, for each subject, in a maximum number of dataset. Also see https://github.com/ivadomed/ivadomed/issues/471.

**Comment by charleygros** [comment]
>     * Check ground truth segmentations
>       
>       * Especially: update the ones for the T1w contrast (under-segmented).
>       * We adopt a "deep active learning" strategy: start training a "good" model with little data, then apply it to a bigger pool, then correct segmentations in a few data, then re-train with bigger pool, etc.
> 
>     * Establish a procedure for co-registering all existing contrasts, for each subject, in a maximum number of dataset. Also see #471.

A suggestion of "First step" would be to (maybe before re doing the GT / co registration):
- train a basic 2D Unet on `sct_testing/large` + `spine-generic` with "Pool all the contrasts" strategy
- run this model on an independent, heterogeneous and challenging dataset made of images with lesions, severe compressions and exotic contrasts
- identify the flaws of this model and iteratively brainstorm solutions for each of them: eg (1) T1w undersegmentation --> manually correct T1w GT ; (2) difference of performance across contrasts --> let's help the network with contrast info via FiLM... helpful? if not, train per contrast? etc. ; (3) improvement needed in cord delineation --> use a soft training ; (4) CSA measures vary across contrast --> use co registered GT during training when possible.

**Comment by mpompolas** [comment]
thanks @jcohenadad.

I have some trouble with VPN access to polymtl. I assume when this is resolved I will also have access to sct_testing/large (not sure where this is located yet).

I had a chat with @andreanne-lemay today and she informed me there is no need to deal with the sct toolbox repo for the training part. I will focus only on Ivadomed, correct?

I'll start with the training on my laptop and when the VPN and access to sct_testing is ready I will move to Rosenberg and follow @charleygros 's suggestions.

One last thing I will probably need:
> * run this model on an independent, heterogeneous and challenging dataset made of images with lesions, severe compressions and exotic contrasts

You have a list of these datasets, or we will have to manually search?

**Comment by jcohenadad** [comment]
> I have some trouble with VPN access to polymtl. I assume when this is resolved I will also have access to sct_testing/large (not sure where this is located yet).

yup! i'll just need you to sign an NDA (i'm working on it)

> I had a chat with @andreanne-lemay today and she informed me there is no need to deal with the sct toolbox repo for the training part. I will focus only on Ivadomed, correct?

correct

> I'll start with the training on my laptop and when the VPN and access to sct_testing is ready I will move to Rosenberg and follow @charleygros 's suggestions.

👍  note that you can also have access to compute canada-- in fact, i would like that the lab start using this amazing ressource-- one collateral benefit is that when `rosenberg` is down, research does not stop 

> One last thing I will probably need:
> 
> > * run this model on an independent, heterogeneous and challenging dataset made of images with lesions, severe compressions and exotic contrasts
> 
> You have a list of these datasets, or we will have to manually search?

it's under `sct_testing/large`. We need to curate it. A meeting will be needed to talk about that. For now there is always the spine-generic dataset which is available

**Comment by mpompolas** [comment]
1.	I tried to start working on the training from my laptop but it runs out of RAM (16GB). Not sure if you want me to spend time trying to optimize the RAM resources during training. Let me know if you want me to dig into this.
2.	Charley sent me a video with instructions on how to use Rosenberg but to my understanding I won’t have access to it until I hear back regarding the NDA and a Grames account and Rosenberg login credentials are issued to me. Through that video I see that I will also need access to the GPU sheet to block GPU resources.
3.	As a last resort I just applied for a compute Canada account with you as my sponsor. My previous accounts are affiliated to my previous labs and are not valid.

**Comment by jcohenadad** [comment]
> I tried to start working on the training from my laptop but it runs out of RAM (16GB). Not sure if you want me to spend time trying to optimize the RAM resources during training. Let me know if you want me to dig into this.

no, that's a waste of time, i think `rosenberg` should be available now

> Charley sent me a video with instructions on how to use Rosenberg but to my understanding I won’t have access to it until I hear back regarding the NDA and a Grames account and Rosenberg login credentials are issued to me. Through that video I see that I will also need access to the GPU sheet to block GPU resources.

NDA: done
GRAME: should be done, if not pls cc me in conversation with JS
GPU sheet: @alexfoias can you pls grant access

> As a last resort I just applied for a compute Canada account with you as my sponsor. My previous accounts are affiliated to my previous labs and are not valid.

i saw and approved 👍

**Comment by jcohenadad** [comment]
I would encourage you to join forces with @leandervaneekelen on https://github.com/ivadomed/ivadomed/issues/471, and especially the next step: https://github.com/ivadomed/ivadomed/issues/471#issuecomment-722006121. Having two people working on this critical pipeline will be very useful. At the end of the day, what matters the most is having high quality training data and ground truth, and this is what this pipeline is about.

**Comment by mpompolas** [comment]
Hi @alexfoias, I can open the GPU sheet, but I can still not edit it. Can you please grant me access?

@jcohenadad 👍

I will finally start working on this this week since my Rosenberg credentials were granted last Friday

**Comment by alexfoias** [comment]
@mpompolas I added you on the google calendar. Sorry for the delay.

**Comment by alexfoias** [comment]
@mpompolas You can not comment/suggest on the Google document.

**Comment by mpompolas** [comment]
Just checked it and it works. Thank you very much @alexfoias

**Comment by mpompolas** [comment]
> * run this model on an independent, heterogeneous and challenging dataset made of images with lesions, severe compressions and exotic contrasts


hey @charleygros @jcohenadad , you have any suggestions for which datasets to test the models on?

**Comment by charleygros** [comment]
> hey @charleygros @jcohenadad , you have any suggestions for which datasets to test the models on?

Yep good question: I feel there are data in `sct_testing/large` which are quite exotic / heterogeneous / challenging!

What I would recommend would be to:
- keep some centers aside, for testing only! For this, there is the param `center_test` in the config file! In particular, I would recommend the "center" `sctUsers`, which is not really a center per se --> data sent by SCT users, often because the current segmentation tools were failing --> Perfect to test the news develloped model, to challenge it!
- keep some patient data in the testing dataset only and look at the model performance specifically for these patients (both quantitatively and qualitatively). Just to name a few: DCM --> `twh`, ALS --> `paris2017Pradat`, MS: `bwh`
- we could also think of a split per disease, using the `method` param in [split config](https://github.com/ivadomed/ivadomed/blob/master/ivadomed/config/config.json#L35), using `pathology` in `participants.tsv`
- to really control the split, we could refine the split "manually" and use the `fname_split` param (here). Not sure if we need to go "[that far](https://github.com/ivadomed/ivadomed/blob/master/ivadomed/config/config.json#L35)", but possible.
- more challenging data? we could ask @lrouhier for high compression data and @andreanne-lemay for SC with tumors.

**Comment by jcohenadad** [comment]
So far, the strategy is the following:
- So far, we focus on simple model (no fancy FILM, etc.) and we compare with `sct_deepseg_sc`
- We need to establish a reproducible framework for comparing performance with `sct_deepseg_sc`
  - This repos would be hosted on a new repos here: https://github.com/sct-pipeline

Action task
- @charleygros identify the list of subjects that were used for `sct_deepseg_sc`
- @lrouhier find the segmentations from the DCM project (AOSpine data)

**Comment by charleygros** [comment]
>     * So far, we focus on simple model (no fancy FILM, etc.) and we compare with `sct_deepseg_sc`
>     * We need to establish a reproducible framework for comparing performance with `sct_deepseg_sc`
>       * This repos would be hosted on a new repos here: https://github.com/sct-pipeline

Agree! With the defaults param of `sct_deepseg_sc` --> output violin plots with Dice!
Would be good to able violin plots for the specifics:
- per contrast_groups (i.e., the ones of `sct_deepseg_sc`, so `t1`, `t2`, `t2s`, `dwi`)
- for specific datasets: subset of spineGeneric, subset of DCM patients (see `pathology` column in `participants.tsv), subset of ALS, subset of MS patients

**Comment by charleygros** [comment]
Suggestion for training/testing splitting:
- Datasets to use: `large` + `spineGeneric`
- Data used to train `sct_deepseg_sc` (I will send this list today) = subset of `large` --> to include into the training dataset of the new model
- New data (i.e., `spineGeneric` + data from `large` which were not used for `sct_deepseg_sc`) --> Do a random split 60% for training validation and 40% for testing. Make sure that the testing includes some MS, ALS, DCM data.

So in summary:
- Training data = training data from `sct_deepseg_sc` + 60% of new data
- Testing data = 40% of new data

@jcohenadad & @mpompolas : What are your thoughts?

**Comment by jcohenadad** [comment]
sounds great. but once we want to ship a model for production, we want to use 100% for training, right?

datasets: i would also add part of UK biobank

**Comment by charleygros** [comment]
> sounds great. but once we want to ship a model for production, we want to use 100% for training, right?

Absolutely! The [above message](https://github.com/ivadomed/ivadomed/issues/495#issuecomment-737566091) aims at drawing a splitting strategy to ensure that the trained model is at least as good as `sct_deepseg_sc` via a fair comparison (i.e., not using old_training_data to do our present testing).

So basically: looking at the results of this comparison:
1. `while (ivadomed_model less good than sct_models)` --> redesign more complicated approach: e.g., hyperparam opt? using FiLM? using one contrast per model?
2. `then` (i.e., ivadomed_model as good or better than sct_models) --> retrain the selected ivadomed using ALL data with available GT

> datasets: i would also add part of UK biobank because these are brain scans. Also we have the manual segs 😁

Oh great, updating [above message](https://github.com/ivadomed/ivadomed/issues/495#issuecomment-737566091).

**Comment by charleygros** [comment]
@mpompolas : Here the dataset used to train `sct_deepseg_sc`:
`/home/GRAMES.POLYMTL.CA/chgroc/duke/projects/ms_seg/sct_deepseg_sc_info/dataset.pkl`

This pandas contains a column named `train_valid_test` with 1 for train, 2 for validing, 3 for testing.

Note: the subject names are slightly different from the one in `large` because it has been done before we moved to the BIDS convention! Please let me know (or @alexfoias ?) if you need some help with that!

**Comment by jcohenadad** [comment]
> Here the dataset used to train sct_deepseg_sc:
duke/projects/ms_seg/sct_deepseg_sc_info/dataset.pkl

can we move this to the [deepseg_sc model repos](https://github.com/sct-data/deepseg_sc_models)? and add in the README this relevant information

**Comment by mpompolas** [comment]
this is great. thanks for the thorough explanation.

Two clarifications: 
1. when I checked the spinal cord toolbox I had to indicate the contrast that is being tested (t1, t2, t2s), so I assumed there was a separate model trained for each modality.
Is there also a model that was also trained on all of them that we could compare with a potential "one to rule them all" model?

2. The sc_deep_seg was trained only on the entire or part of the `large` dataset (no spinegenereric)?

**Comment by charleygros** [comment]
>     1. when I checked the spinal cord toolbox I had to indicate the contrast that is being tested (t1, t2, t2s), so I assumed there was a separate model trained for each modality.

Correct, one model per contrast (t1, t2, t2s, dwi). In the above dataframe there is a column called `contrast` which indicates the model_contrast for which this image has been used to, eg `t2s` --> this image has been used to train the `t2s` model.

>        Is there also a model that was also trained on all of them that we could compare with a potential "one to rule them all" model?

We did preliminary experiments where we trained a model with all images together. These preliminary experiments suggested that the model with all contrasts together was less good than the models trained per contrast. So we focused on training models per contrast.
I think we should do the comparison with the models in `sct_deepseg_sc` since there are the ones we want to replace!

>     2. The sc_deep_seg was trained only on the entire or part of the `large` dataset (no spinegenereric)?

Only on a part of `large` (see [here](https://github.com/ivadomed/ivadomed/issues/495#issuecomment-737566091)). `spinegeneric` was not used.

**Comment by charleygros** [comment]
> can we move this to the [deepseg_sc model repos](https://github.com/sct-data/deepseg_sc_models)? and add in the README this relevant information

Done, [here](https://github.com/sct-data/deepseg_sc_models).

**Comment by mpompolas** [comment]
> Done, here.


@charleygros I want to run testing on the pretrained deepseg_sc models you mentioned but I have trouble loading them with keras:

```
from tensorflow.python.keras.models import load_model

file = '/home/nas/PycharmProjects/deepseg_sc_models/t2_sc.h5'
model = keras.models.load_model(file)

```

```

Python 3.8.5 (default, Sep  4 2020, 07:30:14) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.18.1 -- An enhanced Interactive Python. Type '?' for help.
PyDev console: using IPython 7.18.1
Python 3.8.5 (default, Sep  4 2020, 07:30:14) 
[GCC 7.3.0] on linux
In[2]: runfile('/home/nas/PycharmProjects/ivadomed-personal-scripts/convert_sct_models_to_onnx.py', wdir='/home/nas/PycharmProjects/ivadomed-personal-scripts')
2021-01-18 15:26:54.820500: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcudart.so.10.1
Traceback (most recent call last):
  File "/home/nas/.local/lib/python3.8/site-packages/IPython/core/interactiveshell.py", line 3417, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-2-4aed73caf3e8>", line 1, in <module>
    runfile('/home/nas/PycharmProjects/ivadomed-personal-scripts/convert_sct_models_to_onnx.py', wdir='/home/nas/PycharmProjects/ivadomed-personal-scripts')
  File "/home/nas/Programs/pycharm-2020.2.3/plugins/python/helpers/pydev/_pydev_bundle/pydev_umd.py", line 197, in runfile
    pydev_imports.execfile(filename, global_vars, local_vars)  # execute the script
  File "/home/nas/Programs/pycharm-2020.2.3/plugins/python/helpers/pydev/_pydev_imps/_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "/home/nas/PycharmProjects/ivadomed-personal-scripts/convert_sct_models_to_onnx.py", line 15, in <module>
    model = load_model(file)
  File "/home/nas/.local/lib/python3.8/site-packages/tensorflow/python/keras/saving/save.py", line 182, in load_model
    return hdf5_format.load_model_from_hdf5(filepath, custom_objects, compile)
  File "/home/nas/.local/lib/python3.8/site-packages/tensorflow/python/keras/saving/hdf5_format.py", line 177, in load_model_from_hdf5
    model = model_config_lib.model_from_config(model_config,
  File "/home/nas/.local/lib/python3.8/site-packages/tensorflow/python/keras/saving/model_config.py", line 55, in model_from_config
    return deserialize(config, custom_objects=custom_objects)
  File "/home/nas/.local/lib/python3.8/site-packages/tensorflow/python/keras/layers/serialization.py", line 171, in deserialize
    return generic_utils.deserialize_keras_object(
  File "/home/nas/.local/lib/python3.8/site-packages/tensorflow/python/keras/utils/generic_utils.py", line 354, in deserialize_keras_object
    return cls.from_config(
  File "/home/nas/.local/lib/python3.8/site-packages/tensorflow/python/keras/engine/training.py", line 2238, in from_config
    return functional.Functional.from_config(
  File "/home/nas/.local/lib/python3.8/site-packages/tensorflow/python/keras/engine/functional.py", line 616, in from_config


```


Is there another package or command you use to load it?

I can load it with h5py so the files are not corrupt.

**Comment by charleygros** [comment]
@mpompolas : if you use the SCT python it should work (as `sct_deepseg_sc` does). Or directly using `sct_deepseg_sc` and batch script? Would it be useful?

**Comment by mpompolas** [comment]
Thanks @charleygros. Doing the loading through SCT should work. 
I just want to load the SCT trained models and convert them to .onnx or .pt so I can do a comparison with the newly trained models.

**Comment by charleygros** [comment]
> Thanks @charleygros. Doing the loading through SCT should work.

Yes, I think so too!

>  convert them to .onnx or .pt so I can do a comparison with the newly trained models.

Why do you need to convert them to this format? (Not sure if that's possible (I never tried it tho).)

**Comment by mpompolas** [comment]
> Why do you need to convert them to this format? (Not sure if that's possible (I never tried it tho).)

Lazy way to use the ivadomed toolbox for the training with the pretrained SCT models (check issue [643](https://github.com/ivadomed/ivadomed/issues/643)) 

I thought of converting the trained SCT model to "~/logs/best_model.pt" and run the testing. This would create automatically a ~/logs/results/evalueation3Dmetrics.csv and easily do the the violinplots comparisons we discussed about previously.


Since this seems to be more complicated than I thought, I will just use sct_deepseg on the testing set and manually create the required files.

**Comment by charleygros** [comment]
> I thought of converting the trained SCT model to "~/logs/best_model.pt" and run the testing. This would create automatically a ~/logs/results/evalueation3Dmetrics.csv and easily do the the violinplots comparisons we discussed about previously.

OoooOOoooo I see! That was clever!!

> Since this seems to be more complicated than I thought, I will just use sct_deepseg on the testing set and manually create the required files.

Ok, makes sense to me!
